Improve ability to find reports

ThisIsMissEm opened the pull request:
#### Fixes #31234 and #22419

This changes the search form to allow searching either reports about or from accounts by domain or `@username` or `@username@domain`, and allows viewing all reports regardless of status. It does not allow for search reports from `@username@domain`, since we don't know the username for remote reports.

The "Origin of reported account" options only really make sense if you're not currently searching (since searching implies an origin), so if you search, then this resets to all, if you select an option, search resets.

<img width="3456" height="1942" alt="image" src="https://github.com/user-attachments/assets/89f51810-89bd-4c80-98ed-baba99e65ff1" />

You can also search by a given account, by using the `@username` or `@username@domain.example` syntax. If you omit the domain, then it assumes you mean a local account, if you include the local domain, then it works as expected.

<img width="3456" height="1942" alt="image" src="https://github.com/user-attachments/assets/feebe40c-bdac-41c5-80b8-dfc5cf9d938d" />

<img width="3456" height="1942" alt="image" src="https://github.com/user-attachments/assets/8870f797-6833-4092-b429-cc7fd0c4516c" />

<img width="3456" height="1942" alt="image" src="https://github.com/user-attachments/assets/73b85691-a497-4132-855d-e8ebd1fcf39c" />

When the domain search is "invalid":
<img width="3456" height="1942" alt="image" src="https://github.com/user-attachments/assets/40a5b592-8a48-4830-8013-51c5d6cb59b3" />




I've also added an empty state for when no reports match the current filter: 

<img width="894" alt="Screenshot 2024-08-16 at 03 07 54" src="https://github.com/user-attachments/assets/c0d5055f-2c5a-4f44-9119-a6f96cda6a62">

ThisIsMissEm commented:
One issue I do have is that the rows for each report do not indicate the report status (since "all" was never an option), so we probably need to find a way to indicate that a report has been resolved in the list view.

The click targets for reports are also kinda annoying (due to the grouping)

ThisIsMissEm commented:
I should possibly also use a select instead of the radio options, and then have the select side-by-side with the search term input field.

```
[ search type select ] [ search term input ]
```

However, this would need correct styling for a `<select>` that looks similar-ish to the `<input>`

ThisIsMissEm commented:
Okay, have switched to a select, which looks much nicer.

ThisIsMissEm commented on app/models/report_filter.rb:
```diff
@@ -5,38 +5,67 @@ class ReportFilter
   def initialize(params)
     @params = params
   end
 
   def results
-    scope = Report.unresolved
+    scope = if params[:resolved] == '-1'
+              Report.unscoped
+            elsif params[:resolved] == '1'
+              Report.resolved
+            else
+              Report.unresolved
+            end
```
Should probably change this to `params[:status]`, and then add a redirect that replaces `resolved=1` with `status=resolved`, since the `-1` is absolutely a hack here.

ThisIsMissEm replied:
Have done this.

ThisIsMissEm commented on app/models/report_filter.rb:
```diff
@@ -5,38 +5,67 @@ class ReportFilter
     resolved
     account_id
     target_account_id
```
In theory we could replace these with search_type and search_term (where the term is the `Account#acct`), but for now I think I'll leave these as is, since that requires changes elsewhere.

ThisIsMissEm replied:
I'll probably do up a follow up PR to change those inbound links to the reports list page to use the search capability instead, and deprecate `account_id` and `target_account_id` as parameters.

ThisIsMissEm commented on app/models/report_filter.rb:
```diff
@@ -5,38 +5,67 @@ class ReportFilter
+
   def scope_for(key, value)
     case key.to_sym
-    when :by_target_domain
-      Report.where(target_account: Account.where(domain: value))
```
This is retained by a redirect in reports_controller (though we may need to handle additional query params there)

ThisIsMissEm commented on app/controllers/admin/reports_controller.rb:
```diff
@@ -6,6 +6,17 @@ class ReportsController < BaseController
+      # We previously only supported searching by target account domain for
+      # reports, we now have more search options, but it's important that we
+      # don't break any saved queries people may have:
+      if params.include? :by_target_domain
+        redirect_to admin_reports_path({
+          search_type: 'target',
+          search_term: params[:by_target_domain],
+        })
+      end
```
This doesn't keep the report status filter, but probably should

ThisIsMissEm commented on app/views/admin/reports/index.html.haml:
```diff
@@ -5,31 +5,34 @@
-                          value: params[key],
-                          class: 'string optional',
-                          placeholder: I18n.t("admin.reports.#{key}")
+    .input.string.optional.input--with-select
```
I'm not really sure what styling `.input` `.string` or .optional` actually applies here?

ThisIsMissEm replied:
I think `.optional` and `.string` are only used if there's a label present? I'm not sure.

ThisIsMissEm commented on app/models/report_filter.rb:
```diff
@@ -5,38 +5,67 @@ class ReportFilter
     scope
   end
 
+  def search_filter
```
We probably need input validation here, so checking that the input is either a well-formed domain, or a `@username` or `@username@domain.example`

Currently if you enter say `@username@123` it just returns empty results.

ThisIsMissEm replied:
I tried using the `AccountSearchService::MENTION_ONLY_RE` to raise an error, but that seemingly doesn't error on `@username@123`, but it will error on say `@username 123`

ThisIsMissEm commented on app/controllers/admin/reports_controller.rb:
```diff
@@ -57,6 +63,21 @@ def filter_params
+        search_term: by_target_domain,
+        status: resolved == '1' ? 'resolved' : nil,
+      })
+    end
```
@mjankowski hey, is there a "better" way to do something like this?

The assumption is that moderators likely have links bookmarked to the previously valid filters that have been replaced by search and status filters, so we should keep those filters working by redirecting.

mjankowski replied:
I think having the request specs here (like you do) and then either something like this in the controller, or something in the routes (which detects the old style requests and redirects them from there) are both fine.

Maybe a TODO comment somewhere in here to remove the redirect after some reasonable amount of time or hard version bump or something.

ThisIsMissEm replied:
I'm not sure when / if we'd hard deprecate. Maybe a "# TODO: Delete for 5.0" or something?

ThisIsMissEm commented:
As ReportFilter is reused between the UI and API, we need to figure out what to do with [`Api::V1::Admin::ReportsController`](https://github.com/mastodon/mastodon/blob/b5115850bb22a3676456732b61408fba07bb4b64/app/controllers/api/v1/admin/reports_controller.rb). Should it still accept `resolved` as a param? Why does it re-declare `FILTER_PARAMS` instead of using `ReportFilter::KEYS`?

ThisIsMissEm commented on app/views/admin/reports/index.html.haml:
```diff
@@ -5,31 +5,34 @@
-        = form.text_field key,
-                          value: params[key],
-                          class: 'string optional',
-                          placeholder: I18n.t("admin.reports.#{key}")
```
By removing this line, the following localisation keys became unused:

```
+--------+----------------------------------------------+-------------------------------------------------------------+
| Locale | Key                                          | Value                                                       |
+--------+----------------------------------------------+-------------------------------------------------------------+
|   en   | admin.reports.quick_actions_description_html | Take a quick action or scroll down to see reported content: |
|   en   | admin.reports.reopen                         | Reopen report                                               |
|   en   | admin.reports.reported_account               | Reported account                                            |
|   en   | admin.reports.updated_at                     | Updated                                                     |
+--------+----------------------------------------------+-------------------------------------------------------------+
```

I don't know how to resolve this.

ThisIsMissEm replied:
Have just gone ahead and removed these — hopefully they're not used somehow

ThisIsMissEm commented:
I've made the `status` parameter required, such that we always correctly highlight the "report status filter" menu selector:

<img width="224" alt="Screenshot 2024-08-16 at 23 24 55" src="https://github.com/user-attachments/assets/4f3ac6d2-c733-4e85-91f1-cce199508d34">

This was a bit awkward due to how `filter_link_to` works — it doesn't have a way to say "hey, this is the default option"

This does involve some more redirects, but gives a more understandable user experience

ThisIsMissEm commented:
Went through some user testing & research with @quintessence to help find anything that didn't quite make sense in this new functionality:

- we changed the search to always switch to "all" report statuses, but you can still filter down to unresolved or resolved.
- we added the report status to the rows, above the assignee (previously appearing often as `-` in the UI, but now displays description text of "Not Assigned" if no one is assigned)

ThisIsMissEm commented on app/controllers/admin/reports_controller.rb:
```diff
@@ -6,6 +6,22 @@ class ReportsController < BaseController
+      # this ensures the "status" option menu always shows a highlighted option.
+      if filter_params.exclude? :status
+        if params.slice(*ReportFilter::DIRECT_KEYS).present?
+          return redirect_to admin_reports_path(filter_params.merge({ status: 'all' }))
```
Previously if you went to a specific Account then either of "Reports by others" or "Made reports", the results shown would be "unresolved" reports, not all the reports like you'd expect.

<img width="558" alt="Screenshot 2024-08-16 at 23 32 38" src="https://github.com/user-attachments/assets/7c9eaf0b-5482-4ffb-be21-812c380a4d48">

ThisIsMissEm replied:
However, for `target_origin` without status, you probably want "unresolved" reports.

ThisIsMissEm commented:
@mattly wow, thanks for the in-depth feedback.

You're actually the second person to mention the size of the search & reset buttons, and whilst I agree that they take a lot of space, this is consistent with other filtering forms in the moderation UI, so out of scope for change

To answer these:

- How does the results page handle a lot of matching items?

The results paginates, but is grouped by report subject.

- Are these grouped somehow? The visual hierarchy in the screenshot implies to me they could be grouped by report subject, but since there's no example of having multiple reports by subject, that's just a guess

Yes, see above. As for multiple reports, you get multiple rows:

<img width="842" alt="Screenshot 2024-08-16 at 23 41 05" src="https://github.com/user-attachments/assets/c63c76d5-e5ba-44ad-b775-0ed020c393b7">


- Does this coalesce multiple reports about the same post together into the same row, or will those be displayed individually?

The individual reports within a given report subject are not grouped at present, since report content may vary.

The longer reason why not is because you're not actually reporting the posts independently, you're reporting the account and the statuses are extra information, so you can have multiple reports about the same account, but those reports possibly vary with their contents. 

Ideally we support reporting a status or an account, or another entity, see #30511 for maybe more information.

- How are these sorted? Should that sorting mechanism be exposed in the search interface?

Chronologically. We could add an oldest / newest select.

ThisIsMissEm commented:
@mattly so like this is the hashtag filtering and account filtering:
<img width="866" alt="Screenshot 2024-08-16 at 23 49 54" src="https://github.com/user-attachments/assets/58c7f56b-778b-4783-bd92-ff9571b930ef">


<img width="857" alt="Screenshot 2024-08-16 at 23 50 13" src="https://github.com/user-attachments/assets/14948ac1-26f4-4fc7-96ba-c7226e2c602c">

ThisIsMissEm commented:
> I'm also a bit unclear about the structure of the result list items: it occurs to me that in the first example, Kris is report sender and prudence0 is the author of the reported post, buuuut my initial didn't-think-too-much-about it reading of it had those inverted: I read it as Kris being the author of the "You nasty" post and prudence0 as the reporter. I think there are things which could be done to make these roles clearer, as well as pulling in similar notes/information about both parties.

@mattly oh, also, "you nasty" is the comment on the report, not the post text (because once again, you're reporting the account, not the statuses). Kris is the target of the report in this case, so it is "inverted" as you say.

I think the overall "show a list of reports" UI needs to be redone, but that's out of scope for this PR.

ThisIsMissEm commented:
@mattly had another crack at the list items, taking onboard you feedback, over in https://github.com/mastodon/mastodon/pull/31557

ThisIsMissEm commented on app/controllers/api/v1/admin/reports_controller.rb:
```diff
@@ -14,16 +14,13 @@ class Api::V1::Admin::ReportsController < Api::BaseController
   def index
     authorize :report, :index?
+
+    return redirect_to api_v1_admin_reports_path filter_params.merge(status: 'resolved') if params[:resolved].present?
```
I think a redirect here is what makes most sense, since many http clients will automatically follow redirects, allowing them to continue working.

ThisIsMissEm replied:
Since we didn't originally support search here, we don't need to support quite so many redirects.

ThisIsMissEm commented on app/controllers/api/v1/admin/reports_controller.rb:
```diff
@@ -14,16 +14,13 @@ class Api::V1::Admin::ReportsController < Api::BaseController
-  ).freeze
-
-  PAGINATION_PARAMS = (%i(limit) + FILTER_PARAMS).freeze
+  PAGINATION_PARAMS = (%i(limit) + ReportFilter::KEYS).freeze
```
This expands the available filtering to all the supported ReportFilter options.

ThisIsMissEm commented:
Looks like I've a test case failing, will need to look into that.

mjankowski commented:
I see a review requested from me on this one -- apologies, must have missed it initially!

That said, without getting too into detail yet, I suspect that...

- Get this a code rebase to fix conflicts
- As a side effect, we'll get a CI rebase which will get current ruby version matrix and (hopefully) remove some of annotation noise?
- The CI logs are gone now, but it looks like there may have been some actual failures here?
- Maybe deal w/ that one codeql thing (or comment about how we can ignore it)

Happy to review at that point ... though I do question on this one like so many others if anyone is going to have bandwidth to review/merge anything that's more than one small change at a time.

ThisIsMissEm committed:
Improve filtering of reports

ThisIsMissEm committed:
Improve redirection

ThisIsMissEm committed:
Remove outdated filters

ThisIsMissEm committed:
Update comment

ThisIsMissEm committed:
Remove debugs

ThisIsMissEm committed:
Preserve old API interface

ThisIsMissEm committed:
Update specs

ThisIsMissEm committed:
Update specs
